File: Samples CSharp/AddIns/DocumentEvents/DocEventAddIn.cs
using System;
using System.Collections.Generic;
using System.IO;
using System.Runtime.InteropServices;
using System.Text;
using System.Windows.Forms;
using Therefore.API;

namespace AddInSamples
{
    [Guid("9DDC5D67-A86F-4b57-A862-C75F7F51B124"), ComVisible(true), ClassInterface(ClassInterfaceType.None)]
    public class DocEventAddIn : ITheAddIn
    {
        public void GetHandledEvents(TheClientType client, TheEventSet eventSet)
        {
            eventSet.Add(TheEventType.RetrieveDocument);
            eventSet.Add(TheEventType.DocumentRetrieved);
        }

        public int HandleEvent(TheEvent e)
        {
            // RetrieveDocument occurs BEFORE the document is being retrieved
            // Ask the user if he/she really wants to retrieve it
            if (e.EventType == TheEventType.RetrieveDocument)
            {
                DialogResult result = MessageBox.Show("Do you really want to retrieve the document from Therefore™?", "Retrieve Document", MessageBoxButtons.YesNo);
                if (result == DialogResult.Yes)
                    return 0;
                else if (result == DialogResult.No)
                    return 1;
            }
            // DocumentRetrieved occurs AFTER the document has been retrieved
            // Display a success message
            else if (e.EventType == TheEventType.DocumentRetrieved)
            {
                MessageBox.Show("Document successfully retrieved from Therefore™.");
            }

            return 0;
        }
    }
}

File: Samples VB.NET/AddIns/DocumentEvents/DocEventAddIn.vb
Imports System.IO
Imports System.Runtime.InteropServices
Imports System.Text
Imports Therefore.API

Namespace AddInSamples
	<Guid("9DDC5D67-A86F-4b57-A862-C75F7F51B124"), ComVisible(True), ClassInterface(ClassInterfaceType.None)> _
	Public Class DocEventAddIn
		Implements ITheAddIn
		Public Sub GetHandledEvents(ByVal client As TheClientType, ByVal eventSet As TheEventSet) Implements ITheAddIn.GetHandledEvents
			eventSet.Add(TheEventType.CheckOut)
			eventSet.Add(TheEventType.CheckIn)
		End Sub

		Public Function HandleEvent(ByVal e As TheEvent) As Integer Implements ITheAddIn.HandleEvent
			' Ask before checkout
			If e.EventType = TheEventType.CheckOut Then
				Dim result As DialogResult = MessageBox.Show("Do you really want to check out the document from Therefore™?", "Check out Document", MessageBoxButtons.YesNo)
				If result = DialogResult.Yes Then
					Return 0
				ElseIf result = DialogResult.No Then
					Return 1
				End If
			' check in notification
			ElseIf e.EventType = TheEventType.CheckIn Then
				MessageBox.Show("Document has been changed. Checking document in.")
			End If
			Return 0
		End Function
	End Class
End Namespace
